feat(curriculum): add typescript review page (#59548)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/25-front-end-development/review-typescript/6724e296dceca21b82426229.md

Review the concepts below to prepare for the upcoming quiz.

## What is TypeScript

> - **JavaScript**: JavaScript is a dynamically-typed language. This means that variables can receive any values at run time. The challenge of a dynamically-typed language is that the lack of type safety can introduce errors. 

Reviewer: let's add an example here for reminder.
something like `let data = some random string;` then change it to `data = some random object or whatever`. 
Just to drive the point home further

- **TypeScript**: TypeScript extends the JavaScript language to include static typing, which helps catch errors caused by type mismatches before you run your code.
- **Compiler**: You first need to compile TypeScript code into regular JavaScript. When you run the compiler, TypeScript will evaluate your code and throw an error for any issues where the types don't match.

For example, you can define a type for the `array` parameter as follows:

```typescript
const getRandomValue = (array: string[]) => {
  return array[Math.floor(Math.random() * array.length)];
}
```

> This type definition tells TypeScript that the `array` argument must be an array of strings. Then, when you call `getRandomValue` and pass it a number, you get an error called a compiler error.

Reviewer: ```suggestion
This type definition tells TypeScript that the `array` parameter must be an array of strings. Then, when you call `getRandomValue` and pass it a number, you get an error called a compiler error.
```

  - **`any`**: `any` indicates that a value can have any type. It tells the compiler to stop caring about the type of that variable.
  - **`unknown`**: `unknown` tells TypeScript that you _do_ care about the type of the value, but you don’t actually know what it is. `unknown` is generally preferred over `any`.
  - **`void`**: This is a special type that you’ll typically only use when defining functions. Functions that don’t have a return value use a return type of `void`.
  - **`never`**: It represents a type that will never exist.

- **`type` Keyword**: This keyword is like `const`, but instead of declaring a variable, you can declare a type.

It is useful for declaring custom types such as union types or types that include only specific values:

```typescript
type stringOrNumber = string | number;
type bot = "camperchan" | "camperbot" | "naomi";
```

- **`interface`**: Interfaces are like classes for types. They can implement or extend other interfaces, are specifically object types, and are generally preferred unless you need a specific feature offered by a `type` declaration.

```typescript
interface wowie {
  zowie: boolean;
  method: () => void;
}
```

> - **Defining Return Type**: You can also define the _return type_ of the function.

Reviewer: let's add a sentence here explaining that in the example, the function return must be a a string otherwise TypeScript will show an error message.

```typescript
const getRandomValue = (array: string[]): string => {
  return array[Math.floor(Math.random() * array.length)];
}
```

> 

Reviewer: ```suggestion
```

## Generics

- **Defining Generic Type**: You can define a generic type and refer to it in your function. It can be thought of as a special parameter you provide to a function to control the behavior of the function's type definition.

Here is an example of defining a generic type for a function:

```typescript
const getRandomValue = <T>(array: T[]): T => {
  return array[Math.floor(Math.random() * array.length)];
}
const val = getRandomValue([1, 2, 4])
```

> The `<T>` syntax tells TypeScript that you are defining a generic type `T` for the function. T is a common name for generic types, but you can use any name.

Reviewer: ```suggestion
The `<T>` syntax tells TypeScript that you are defining a generic type `T` for the function. `T` is a common name for generic types, but you can use any name.
```

```typescript
const email = document.querySelector("#email");

if (email instanceof HTMLInputElement) {
    console.log(email.value);
}
```

- **Casting**: When TypeScript cannot automatically determine the type of a value, such as the result from `request.json()` method in the example below, you'll get a compiler error. One way to resolve this is by casting the type, but doing so weakens TypeScript's ability to catch potential errors.

```typescript
interface User {
    name: string;
    age: number;
}

const printAge = (user: User) => 
  console.log(`${user.name} is ${user.age} years old!`)

const request = await fetch("url")
const myUser = await request.json() as User;
printAge(myUser);
```

- **Type Guard**: Instead of casting the type, you can write a type guard:

```typescript
interface User {
    name: string;
    age: number;
}

const isValidUser = (user: unknown): user is User => {
  return !!user && 
    typeof user === "object" &&
    "name" in user &&
    "age" in user;
}
```

The `user is User` syntax indicates that your function returns a boolean value, which when true means the `user` value satisfies the `User` interface.

## `tsconfig` File

> - **`tsconfig.json`**: TypeScript’s compiler settings live in a `tsconfig.json` file in the root directory of your project.

Reviewer: ```suggestion
- **`tsconfig.json`**: TypeScript's compiler settings live in a `tsconfig.json` file in the root directory of your project.
```
